### ISSUE TITLE
[web] Navigator 2.0 back/forward button reporting breaks on dart >= 12.0-0

### ISSUE THREAD

**Issue by lulupointu** [flutter/flutter#75489]
<details>
  <summary>Here is a small reproducible example</summary>

```dart
import 'package:flutter/foundation.dart';
import 'package:flutter/material.dart';

void main() {
  runApp(UrlHandler());
}

class UrlHandler extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp.router(
      routerDelegate: UrlHandlerRouterDelegate(),
      routeInformationParser: UrlHandlerInformationParser(),
    );
  }
}

class NavigationState {
  final int value;
  NavigationState(this.value);
}

final GlobalKey<NavigatorState> _urlHandlerRouterDelegateNavigatorKey =
GlobalKey<NavigatorState>();
class UrlHandlerRouterDelegate extends RouterDelegate<NavigationState>
    with ChangeNotifier, PopNavigatorRouterDelegateMixin {
  int count = 0;

  @override
  Widget build(BuildContext context) {
    return Navigator(
      pages: [
        MaterialPage(child: MyHomePage(count: count, increase: increase)),
      ],
      onPopPage: (_, __) {
        // We don't handle routing logic here, so we just return false
        return false;
      },
    );
  }

  @override
  GlobalKey<NavigatorState> get navigatorKey => _urlHandlerRouterDelegateNavigatorKey;

  // Navigation state to app state
  @override
  Future<void> setNewRoutePath(NavigationState navigationState) async {
    // If a value which is not a number has been entered,
    // navigationState.value is null so we just notifyListeners
    // without changing the app state to change the value of the url
    // to its previous value
    if (navigationState.value == null) {
      notifyListeners();
      return null;
    }

    // Get the new count, which is navigationState.value//10
    count = (navigationState.value / 10).floor();

    // If the navigationState.value was not a multiple of 10
    // the url is not equal to count*10, therefore the url isn't right
    // In that case, we notifyListener in order to get the valid NavigationState
    // from the new app state
    if (count * 10 != navigationState.value) notifyListeners();
    return null;
  }

  // App state to Navigation state, triggered by notifyListeners()
  @override
  NavigationState get currentConfiguration => NavigationState(count*10);

  void increase() {
    count++;
    notifyListeners();
  }
}

class UrlHandlerInformationParser extends RouteInformationParser<NavigationState> {
  // Url to navigation state
  @override
  Future<NavigationState> parseRouteInformation(RouteInformation routeInformation) async {
    print('Got new route information with state: ${routeInformation.state}');
    return NavigationState(int.tryParse(routeInformation.location ?? '') ?? 0);
  }

  // Navigation state to url
  @override
  RouteInformation restoreRouteInformation(NavigationState navigationState) {
    print('restoreRouteInformation with location "${'/${navigationState.value}'}"');
    return RouteInformation(location: '/${navigationState.value}');
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, @required this.count, @required this.increase}) : super(key: key);
  final int count;
  final VoidCallback increase;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('Counter App'),
      ),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            Text(
              'You have pushed the button this many times:',
            ),
            Text(
              '${widget.count}',
              style: Theme.of(context).textTheme.headline4,
            ),
          ],
        ),
      ),
      floatingActionButton: FloatingActionButton(
        onPressed: () {
          widget.increase();
        },
        tooltip: 'Counter',
        child: Icon(Icons.add),
      ),
    );
  }
}
```

</details>

### How to reproduce the bug
1. Launch the app
2. Tap the fab, this will push a new RouteInformation
3. Tap the backward button

### What happens
`parseRouteInformation` from the `RouteInformationParser` should be triggered but it is not. We can see that the `print` message inside it is not displayed

### Here is a video footage of the bug

https://user-images.githubusercontent.com/5128921/107014430-e1c04c80-679b-11eb-8843-c6ece323ed39.mp4

#### This bug appears only with sdk: ">=2.12.0-0 <3.0.0".
Please try to run this app with `sdk: ">=2.11.0-0 <3.0.0"` (without null safety) for example, and you will see no issues
#### This bug is fixed if we report a state in RouteInformation
Here is the same code, but along with the RouteInformation location, I report a state
<details>
  <summary>The example</summary>

```dart
import 'package:flutter/foundation.dart';
import 'package:flutter/material.dart';
import 'dart:html' as html;

void main() {
  runApp(UrlHandler());
}

class UrlHandler extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp.router(
      routerDelegate: UrlHandlerRouterDelegate(),
      routeInformationParser: UrlHandlerInformationParser(),
    );
  }
}

class NavigationState {
  final int? value;
  NavigationState(this.value);
}

final GlobalKey<NavigatorState> _urlHandlerRouterDelegateNavigatorKey =
GlobalKey<NavigatorState>();
class UrlHandlerRouterDelegate extends RouterDelegate<NavigationState>
    with ChangeNotifier, PopNavigatorRouterDelegateMixin {
  int count = 0;

  @override
  Widget build(BuildContext context) {
    return Navigator(
      pages: [
        MaterialPage(child: MyHomePage(count: count, increase: increase)),
      ],
      onPopPage: (_, __) {
        // We don't handle routing logic here, so we just return false
        return false;
      },
    );
  }

  @override
  GlobalKey<NavigatorState> get navigatorKey => _urlHandlerRouterDelegateNavigatorKey;

  // Navigation state to app state
  @override
  Future<void> setNewRoutePath(NavigationState navigationState) async {
    // If a value which is not a number has been entered,
    // navigationState.value is null so we just notifyListeners
    // without changing the app state to change the value of the url
    // to its previous value
    if (navigationState.value == null) {
      notifyListeners();
      return null;
    }

    // Get the new count, which is navigationState.value//10
    count = (navigationState.value! / 10).floor();

    // If the navigationState.value was not a multiple of 10
    // the url is not equal to count*10, therefore the url isn't right
    // In that case, we notifyListener in order to get the valid NavigationState
    // from the new app state
    if (count * 10 != navigationState.value) notifyListeners();
    return null;
  }

  // App state to Navigation state, triggered by notifyListeners()
  @override
  NavigationState get currentConfiguration => NavigationState(count*10);

  void increase() {
    count++;
    notifyListeners();
  }
}

class UrlHandlerInformationParser extends RouteInformationParser<NavigationState> {
  // Url to navigation state
  @override
  Future<NavigationState> parseRouteInformation(RouteInformation routeInformation) async {
    print('Got new route information with state: ${routeInformation.state}');
    return NavigationState(int.tryParse(routeInformation.location ?? '') ?? 0);
  }

  // Navigation state to url
  @override
  RouteInformation restoreRouteInformation(NavigationState navigationState) {
    print('restoreRouteInformation with location "${'/${navigationState.value}'}"');
    return RouteInformation(location: '/${navigationState.value}', state: navigationState.value);
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key? key, required this.count, required this.increase}) : super(key: key);
  final int count;
  final VoidCallback increase;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('Counter App'),
      ),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            Text(
              'You have pushed the button this many times:',
            ),
            Text(
              '${widget.count}',
              style: Theme.of(context).textTheme.headline4,
            ),
          ],
        ),
      ),
      floatingActionButton: FloatingActionButton(
        onPressed: () {
          widget.increase();
        },
        tooltip: 'Counter',
        child: Icon(Icons.add),
      ),
    );
  }
}
```
</details>
Reproduce the same steps as before: the behaviour is fixed!

### Logs of the buggy app
<details>
  <summary>Logs</summary>

```
flutter run --verbose
[  +52 ms] executing: [/usr/lib/flutter/] git -c log.showSignature=false log -n 1 --pretty=format:%H
[  +30 ms] Exit code 0 from: git -c log.showSignature=false log -n 1 --pretty=format:%H
[   +1 ms] 5d36f2e7f5387b6c751449258ade8e4e6edf99be
[        ] executing: [/usr/lib/flutter/] git tag --points-at 5d36f2e7f5387b6c751449258ade8e4e6edf99be
[  +45 ms] Exit code 0 from: git tag --points-at 5d36f2e7f5387b6c751449258ade8e4e6edf99be
[        ] 1.25.0-8.3.pre
[  +36 ms] executing: [/usr/lib/flutter/] git rev-parse --abbrev-ref --symbolic @{u}
[   +4 ms] Exit code 0 from: git rev-parse --abbrev-ref --symbolic @{u}
[        ] origin/beta
[        ] executing: [/usr/lib/flutter/] git ls-remote --get-url origin
[   +3 ms] Exit code 0 from: git ls-remote --get-url origin
[        ] https://github.com/flutter/flutter.git
[  +28 ms] executing: [/usr/lib/flutter/] git rev-parse --abbrev-ref HEAD
[   +5 ms] Exit code 0 from: git rev-parse --abbrev-ref HEAD
[        ] beta
[  +31 ms] Artifact Instance of 'AndroidGenSnapshotArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'AndroidInternalBuildArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'IOSEngineArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'FlutterWebSdk' is not required, skipping update.
[   +2 ms] Artifact Instance of 'WindowsEngineArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'MacOSEngineArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'LinuxEngineArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'LinuxFuchsiaSDKArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'MacOSFuchsiaSDKArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'FlutterRunnerSDKArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'FlutterRunnerDebugSymbols' is not required, skipping update.
[  +51 ms] executing: /home/lulupointu/Android/Sdk/platform-tools/adb devices -l
[  +23 ms] List of devices attached
[   +2 ms] Artifact Instance of 'AndroidGenSnapshotArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'AndroidInternalBuildArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'IOSEngineArtifacts' is not required, skipping update.
[   +1 ms] Artifact Instance of 'WindowsEngineArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'MacOSEngineArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'LinuxEngineArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'LinuxFuchsiaSDKArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'MacOSFuchsiaSDKArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'FlutterRunnerSDKArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'FlutterRunnerDebugSymbols' is not required, skipping update.
[  +54 ms] Skipping pub get: version match.
[  +67 ms] Found plugin integration_test at /usr/lib/flutter/packages/integration_test/
[  +83 ms] Found plugin integration_test at /usr/lib/flutter/packages/integration_test/
[   +4 ms] Generating /home/lulupointu/ProgrammingProjects/navigator_2_null_safety/android/app/src/main/java/io/flutter/plugins/GeneratedPluginRegistrant.java
[ +133 ms] Launching lib/main.dart on Chrome in debug mode...
[  +66 ms] Updating assets
[  +73 ms] Waiting for connection from debug service on Chrome...
[  +26 ms] Found plugin integration_test at /usr/lib/flutter/packages/integration_test/
[  +22 ms] Found plugin integration_test at /usr/lib/flutter/packages/integration_test/
[  +42 ms] <- reset
[   +5 ms] /usr/lib/flutter/bin/cache/dart-sdk/bin/dart --disable-dart-dev /usr/lib/flutter/bin/cache/artifacts/engine/linux-x64/frontend_server.dart.snapshot
--sdk-root /usr/lib/flutter/bin/cache/flutter_web_sdk/ --incremental --target=dartdevc --debugger-module-names --experimental-emit-debug-metadata --output-dill
/tmp/flutter_tools.YOQOYL/flutter_tool.DDNOPI/app.dill --libraries-spec file:///usr/lib/flutter/bin/cache/flutter_web_sdk/libraries.json --packages
/home/lulupointu/ProgrammingProjects/navigator_2_null_safety/.dart_tool/package_config.json -Ddart.vm.profile=false -Ddart.vm.product=false --enable-asserts
--track-widget-creation --filesystem-root /tmp/flutter_tools.YOQOYL/flutter_tools.LARBSY --filesystem-scheme org-dartlang-app --initialize-from-dill
build/80b1a4cf4e7b90e1ab5f72022a0bc624.cache.dill.track.dill --platform file:///usr/lib/flutter/bin/cache/flutter_web_sdk/kernel/flutter_ddc_sdk_sound.dill
--sound-null-safety
[   +8 ms] <- compile org-dartlang-app:/web_entrypoint.dart
[+12636 ms] Waiting for connection from debug service on Chrome... (completed in 12.7s)
[        ] Synced 28.3MB.
[        ] <- accept
[        ] Caching compiled dill
[  +95 ms] Using Google Chrome 87.0.4280.141 

[ +547 ms] [CHROME]:
[        ] [CHROME]:DevTools listening on ws://127.0.0.1:37747/devtools/browser/d2b708a4-ecb9-4d31-9a38-b15de5bf6ccc
[ +335 ms] DwdsInjector: Received request for entrypoint at http://localhost:36155/main_module.bootstrap.js
[   +3 ms] MetadataProvider: Loading debug metadata...
[   +6 ms] MetadataProvider: Loaded debug metadata
[   +4 ms] DwdsInjector: Injected debugging metadata for entrypoint at http://localhost:36155/main_module.bootstrap.js
[+1441 ms] DevHandler: Debug service listening on ws://127.0.0.1:41213/r1QvJVx81TI=/ws

[   +5 ms] Debug service listening on ws://127.0.0.1:41213/r1QvJVx81TI=/ws
[        ] 💪 Running with sound null safety 💪
[   +2 ms] Warning: Flutter's support for web development is not stable yet and hasn't
[        ] been thoroughly tested in production environments.
[        ] For more information see https://flutter.dev/web
[        ] 🔥  To hot restart changes while running, press "r" or "R".
[        ] For a more detailed help message, press "h". To quit, press "q".
[ +190 ms] Got new route information with state: null
[   +1 ms] State from browser (html.window.history.state["state"]): null
[ +236 ms] restoreRouteInformation with location "/0" and value "0"
[+23871 ms] restoreRouteInformation with location "/10" and value "10"
[+1148 ms] restoreRouteInformation with location "/20" and value "20"
[ +494 ms] restoreRouteInformation with location "/30" and value "30"
```

```
flutter analyze
Analyzing navigator_2_null_safety...                                    

   info • The library 'package:integration_test/integration_test.dart' is legacy, and should not be imported into a null safe library •
          integration_test/app_test.dart:10:8 • import_of_legacy_library_into_null_safe
   info • The library 'package:integration_test/integration_test_driver.dart' is legacy, and should not be imported into a null safe library •
          integration_test/driver.dart:6:8 • import_of_legacy_library_into_null_safe
   info • Avoid using web-only libraries outside Flutter web plugin packages • lib/main.dart:3:1 • avoid_web_libraries_in_flutter

3 issues found. (ran in 1.4s)
```

```
flutter doctor -v
[✓] Flutter (Channel beta, 1.25.0-8.3.pre, on Linux, locale en_US.UTF-8)
    • Flutter version 1.25.0-8.3.pre at /usr/lib/flutter
    • Framework revision 5d36f2e7f5 (3 weeks ago), 2021-01-14 15:57:49 -0800
    • Engine revision 7a8f8ca02c
    • Dart version 2.12.0 (build 2.12.0-133.7.beta)

[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.2)
    • Android SDK at /home/lulupointu/Android/Sdk
    • Platform android-30, build-tools 30.0.2
    • ANDROID_HOME = /home/lulupointu/Android/Sdk
    • Java binary at: /opt/android-studio/jre/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6222593)
    • All Android licenses accepted.

[✓] Chrome - develop for the web
    • Chrome at google-chrome

[✓] Android Studio
    • Android Studio at /opt/android-studio
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6222593)

[✓] IntelliJ IDEA Ultimate Edition (version 2020.2)
    • IntelliJ at /opt/idea-IU-202.7660.26
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart

[✓] VS Code (version 1.52.1)
    • VS Code at /usr/share/code
    • Flutter extension version 3.19.0

[✓] Connected device (1 available)
    • Chrome (web) • chrome • web-javascript • Google Chrome 87.0.4280.141

• No issues found!
```

</details>

**Comment by pedromassangocode** [comment]
### Running with SDK ">=2.7.0 <3.0.0"

This is my logs running the code in the OP and incrementing counter five times and decrement back five times.

**Comment by lulupointu** [comment]
Wow it seems that it was fix between Flutter v1.25.0-8.3.pre and 1.26.0-17.3.pre since upgrading flutter fixed it !
Sorry for the inconvenience!
